### PR TITLE
[Do not merge!] [Build] Migrate vendored DeepGEMM from pybind to TORCH_LIBRARY (abi3)

### PR DESCRIPTION
**Pull request by cleonard530**
## Purpose
This PR continues the libtorch stable ABI migration (see https://github.com/vllm-project/vllm/issues/26946) for vLLM. Here, we are not migrating deep_gemm to the stable abi yet, but are moving it off of pybind and onto TORCH_LIBRARY. Once we make these updates, we will open a seperate PR to make deep_gemm torch ABI stable. This points to the DeepGEMM updates in the PR https://github.com/deepseek-ai/DeepGEMM/pull/393

- Pin DeepGEMM to `cleonard530/DeepGEMM` (`migrate_pybind_to_torch_library`) so vLLM vendors the TORCH_LIBRARY + abi3 layout instead of pybind11. If this works, we will need to discuss how we want to merge`cleonard530/DeepGEMM`. vLLM has a fork of `DeepGEMM`, but currently the build still points towards upstream.   
- Replace the per-Python `g++` build of `_C.cpython-*.so` with a single `_C_extension.abi3.so` built via DeepGEMM’s `setup.py`, plus `deep_gemm/_C.py` as the `torch.ops.deep_gemm` shim.
- Update cmake / wheel checks / install script for the new artifacts (`_C.py` + `_C_extension*.so`) and drop `DEEPGEMM_PYTHON_INTERPRETERS` / `TARGET_PY`.
This closes the per-interpreter DeepGEMM binding path tracked after the failed single-abi3 attempt (#41476 / #41512) and implements the TORCH_LIBRARY direction noted in #42431.

@janeyx99 @Harry-Chen 

## Test Plan
```
  python tools/check_wheel_deepgemm.py
  pytest -v -s tests/kernels/quantization/test_block_fp8.py
  pytest -v -s tests/kernels/moe/test_deepgemm.py
  pytest -v -s tests/kernels/moe/test_batched_deepgemm.py
  pytest -v -s tests/kernels/attention/test_deepgemm_attention.py
```

## Test Result

- tools/check_wheel_deepgemm.py - `PASS`
- tests/kernels/quantization/test_block_fp8.py - `100 failed, 322 passed`, all failures are `fatal error: sanitizer/asan_interface.h: No such file or directory`. Not related to TORCH_LIBRARY migration
- tests/kernels/moe/test_deepgemm.py - `PASS`
- tests/kernels/moe/test_batched_deepgemm.py - `PASS`
- tests/kernels/attention/test_deepgemm_attention.py - `PASS`

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

Migration progress using the Audit Python extension [torch-abi-audit](https://pypi.org/project/torch-abi-audit/):
Note, deep_gemm used to have `uses-private-api` instead of `abi3-ok`

```
 -- extensions --
    [STABLE  ] [abi3-ok               ] _C_stable_libtorch.abi3.so  (stable_shim=87, unstable=0)
    [STABLE  ] [abi3-ok               ] _flashmla_C.abi3.so  (stable_shim=65, unstable=0)
    [STABLE  ] [abi3-ok               ] _flashmla_extension_C.abi3.so  (stable_shim=63, unstable=0)
    [STABLE  ] [abi3-ok               ] _moe_C_stable_libtorch.abi3.so  (stable_shim=72, unstable=0)
    [NO-TORCH] [abi3-ok               ] cumem_allocator.abi3.so
    [NO-TORCH] [abi3-ok               ] fs_io_C.abi3.so
    [NO-TORCH] [abi3-ok               ] spinloop.abi3.so
--> [UNSTABLE] [abi3-ok               ] third_party/deep_gemm/_C_extension.abi3.so  (stable_shim=0, unstable=71)
    [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa2_C.abi3.so  (stable_shim=0, unstable=85)
    [UNSTABLE] [abi3-ok               ] vllm_flash_attn/_vllm_fa3_C.abi3.so  (stable_shim=0, unstable=81)
```

### CODE REVIEW

File: cmake/external_projects/deepgemm.cmake [hunk 1]
```diff
@@ -87,33 +87,17 @@ if(DEEPGEMM_ARCHS)
...
-  # DEEPGEMM_PYTHON_INTERPRETERS: ":"-separated target Python paths.
-  # Empty/unset → fall back to the build interpreter (editable installs).
-  # (Empty-but-set env vars test as DEFINED in cmake — treat as unset.)
-  if(NOT "$ENV{DEEPGEMM_PYTHON_INTERPRETERS}" STREQUAL "")
-    string(REPLACE ":" ";" _dg_pythons "$ENV{DEEPGEMM_PYTHON_INTERPRETERS}")
-  else()
-    set(_dg_pythons "${Python_EXECUTABLE}")
-  endif()
-  message(STATUS "DeepGEMM _C will be built for: ${_dg_pythons}")
+  message(STATUS "DeepGEMM extension will be built with: ${Python_EXECUTABLE}")
```
Comment (cleonard530) [inline cmake/external_projects/deepgemm.cmake:103]: This used to link private CPython symbols, so each python needed it's own C.cpython-3XY-….so.

Now, it builds against the CPython Limited/Stable ABI (Py_LIMITED_API), so one _C_extension.abi3.so works across supported CPythons.

File: cmake/external_projects/deepgemm.cmake [hunk 2]
```diff
@@ -124,40 +108,29 @@ if(DEEPGEMM_ARCHS)
...
-      COMMENT "Building DeepGEMM _C for ${_pybin}"
-      VERBATIM)
-    list(APPEND _dg_markers "${_dg_marker}")
-    install(DIRECTORY "${_dg_dir}/"
-      DESTINATION vllm/third_party/deep_gemm
-      COMPONENT _deep_gemm_C
-      FILES_MATCHING PATTERN "_C.cpython-*.so")
-  endforeach()
-  add_custom_target(_deep_gemm_C ALL DEPENDS ${_dg_markers})
+  set(_dg_dir "${CMAKE_CURRENT_BINARY_DIR}/deepgemm_C")
```
Comment (cleonard530) [inline cmake/external_projects/deepgemm.cmake:114]: This block replaced the per-Python build loop with a single abi3 build

File: tools/build_deepgemm_C.py [hunk 3]
```diff
@@ -1,85 +1,53 @@
...
+        "expected TORCH_LIBRARY migration layout"
+    )
+
+# Avoid DeepGEMM's clean-git assertion when vendoring a local dirty tree.
+env = os.environ.copy()
+env.setdefault("DG_USE_LOCAL_VERSION", "0")
+env.pop("DG_SKIP_CUDA_BUILD", None)
+
+print(f"[build_deepgemm_C] building in {src} with {sys.executable}", flush=True)
+subprocess.check_call(
```
Comment (cleonard530) [inline tools/build_deepgemm_C.py:38]: Use deep_gemm's setup.py to build deep_gemm

File: tools/check_wheel_deepgemm.py [hunk 4]
```diff
@@ -1,41 +1,45 @@
...
-spec = importlib.util.find_spec("vllm.third_party.deep_gemm")
-if spec is None or spec.origin is None:
-    sys.exit("vllm.third_party.deep_gemm not importable; is vllm installed?")
-pkg_dir = Path(spec.origin).parent
-
-found = {f"{m[1]}.{m[2]}" for f in os.listdir(pkg_dir) if (m := SO_RE.match(f))}
-required = required_pythons()
-missing = [v for v in required if v not in found]
-print(f"deepgemm _C: found {sorted(found)}, required {required}, missing {missing}")
-sys.exit(1 if missing else 0)
```
Comment (cleonard530) [inline tools/check_wheel_deepgemm.py:41]: This file used to check each required python version in `pyproject.toml` and failed if matching *.so python version weren't there.

File: tools/check_wheel_deepgemm.py [hunk 5]
```diff
@@ -1,41 +1,45 @@
...
+    if not shim.is_file():
+        missing.append("_C.py")
+    if not so_files:
+        missing.append("_C_extension*.so")
+
+    print(
+        f"deepgemm vendored binding: shim={shim.is_file()}, "
+        f"extensions={[p.name for p in so_files]}"
+    )
+    if missing:
```
Comment (cleonard530) [inline tools/check_wheel_deepgemm.py:38]: Now, it checks `vllm.third_party.deep_gemm` to make sure the `_C.py` shim and `_C_extension*.so` are there.

**Comment by Harry-Chen** [comment]
Marked as ready to automatically trigger CI run.

**Comment by cleonard530** [comment]
@Harry-Chen, the failure for these three test comes from this check
 [DG_DEVICE_ASSERT((values[j] & 0x807fffffu) == 0); ](https://github.com/cleonard530/DeepGEMM/blob/441c417c6cf7184593421273b7e6d79a0999a8f3/deep_gemm/include/deep_gemm/impls/smxx_layout.cuh#L131-L132)
 
 which was added between the previously tagged commit and my commit, but not directly due to the migration. We will need to figure out how to handle this before anything is merged (I don't think it will be difficult to fix),  but I am planning on putting that off until we decide if we will be using upstream or the [vllm-project/DeepGEMM](https://github.com/vllm-project/DeepGEMM), which doesn't have the check.

**Comment by Harry-Chen** [comment]
> @Harry-Chen, the failure for these three test comes from this check [DG_DEVICE_ASSERT((values[j] & 0x807fffffu) == 0); ](https://github.com/cleonard530/DeepGEMM/blob/441c417c6cf7184593421273b7e6d79a0999a8f3/deep_gemm/include/deep_gemm/impls/smxx_layout.cuh#L131-L132)
> 
> which was added between the previously tagged commit and my commit, but not directly due to the migration. We will need to figure out how to handle this before anything is merged (I don't think it will be difficult to fix), but I am planning on putting that off until we decide if we will be using upstream or the [vllm-project/DeepGEMM](https://github.com/vllm-project/DeepGEMM), which doesn't have the check.

I see. I think currently we are using <https://github.com/deepseek-ai/DeepGEMM/tree/nv_dev>, which adds sm120 support (from NVIDIA) on top of upstream master.

**Comment by cleonard530** [comment]
> I see. I think currently we are using https://github.com/deepseek-ai/DeepGEMM/tree/nv_dev, which adds sm120 support (from NVIDIA) on top of upstream master.

Oh yes, you're right @Harry-Chen. Once we agree we are in a good spot, should we try to merge into this branch? Do you know how receptive the maintainers for this repo will be to our migration efforts? 

CC @janeyx99

File: tools/build_deepgemm_C.py (hunk 3 above)
Comment (cleonard530) [reply tools/build_deepgemm_C.py:38]: This call also gets Python_EXECUTABLE so building DeepGEMM's abi3 extension always tracks whatever Python vLLM is currently being built for.

**Comment by Harry-Chen** [comment]
> > I see. I think currently we are using https://github.com/deepseek-ai/DeepGEMM/tree/nv_dev, which adds sm120 support (from NVIDIA) on top of upstream master.
> 
> Oh yes, you're right @Harry-Chen. Once we agree we are in a good spot, should we try to merge into this branch? Do you know how receptive the maintainers for this repo will be to our migration efforts?

I have an issue there: https://github.com/deepseek-ai/DeepGEMM/issues/333. But given the situation of the repo, I do not think we should put too much hope in upstream PRs. But since NVIDIA has their own branch on master, I do not know whether maintaining a fork would be a good idea.

> CC @janeyx99

File: vllm/model_executor/warmup/deep_gemm_warmup.py [hunk 6]
```diff
@@ -234,7 +224,8 @@ def _deepgemm_fp8_gemm_nt_warmup(
 
     device = w.device
     a1q = torch.empty((max_tokens, k), device=device, dtype=torch.float8_e4m3fn)
-    a1q_scales = torch.empty(
+    # Must be a power of two (UE8M0 packing asserts zero sign/mantissa bits).
+    a1q_scales = torch.ones(
```
Comment (cleonard530) [inline vllm/model_executor/warmup/deep_gemm_warmup.py]: Not related to the TORCH_LIBRARY migration — a separate pre-existing bug. torch.empty left a1q_scales uninitialized, occasionally tripping DeepGEMM's UE8M0 assert (scales must be exact powers of two). Switched to torch.ones; warmup output is discarded, so this has no effect on real inference.

**Comment by cleonard530** [comment]
@Harry-Chen, I opened a PR upstream on the `nv_dev` branch, https://github.com/deepseek-ai/DeepGEMM/pull/393. Please let me know if you think this is the right move and if you know anyone who might be able to review it to help get it through. If there is no response there, then the only other option I can think of is using the https://github.com/vllm-project/DeepGEMM fork, which I know we don't want to have to maintain.

**Comment by cleonard530** [comment]
Hi @Harry-Chen, are you able to rerun the CI for this? I have updated https://github.com/deepseek-ai/DeepGEMM/pull/393 to incorporate the new kernels that were added after I originally opened the PR and want to test if they work in vLLM. Specifically, I have not been able to test them on a Blackwell GPU (sm_100 or sm_120)  yet and I was hoping I could test them through the CI here. The updates to `csrc/apis/sm90_mega.hpp` are similar to the updates to `csrc/apis/mega.hpp` though, and I was able to do a single GPU test on it and they are passing.  

Also, do you have any updates on how we can proceed with this and try and land the DeepGEMM PR so that we can try to migrate it to the torch stable ABI?

**Comment by Harry-Chen** [comment]
@cleonard530 Thank you for your work! I do have some updates:

* As you may have seen, deepgemm has a major update today with many new kernels and updates to existing kernels. And we need these new kernels for deepseek-v4.1 support soon. The new version also incurs other dependencies like deepjit, which needs to be re-evaluated.
* NVIDIA people are shifting their focus, so `nv_dev` branch (with sm120 support and some other customized changes for vllm) may not be updated in a very timely manner. vllm maintainers may take over and maintain a deemgemm fork (just like flashmla).

And for other existing libraries from deepseek after today's update:

* For flashmla, vllm is maintaining its own fork and will cherry-pick upstream changes, still with stable ABI support (https://github.com/vllm-project/FlashMLA/pull/20)
* For deepselect, vllm will fork it and add stable ABI support.

Meanwhile, I and @janeyx99 are still trying our best to push these changes upstream, which should be the most desirable way for all parties.

**Comment by cleonard530** [comment]
@Harry-Chen, Thank you for the update!
